nestkernel/node.h

```diff
@@ -801,6 +814,23 @@ class Node
   virtual double get_tau_syn_ex( int comp );
   virtual double get_tau_syn_in( int comp );
 
+  /**
+   * Compute gradient change for eprop synapses.
+   *
+   * This method is called from an eprop synapse on the eprop target neuron and returns the change in gradient.
```

Comment:
@heplesser @akorgor Do you agree the following versions for both the original and biologically enhanced rules?

```
/**
 * Compute gradient change for eprop synapses.
 *
 * This method is called from an eprop synapse on the eprop target neuron. It updates various parameters related to e-prop plasticity
 * according to Bellec et al. (2020) with additional biological features described in Korcsak-Gorzo, Stapmanns, and Espinoza Valverde
 * et al. (in preparation).
 *
 * @param t_spike [in] Time of the current spike.
 * @param t_spike_previous [in] Time of the previous spike.
 * @param z_previous_buffer [in, out] Value of presynaptic spiking variable from previous time step.
 * @param z_bar [in, out] Filtered presynaptic spiking variable.
 * @param e_bar [in, out] Filtered eligibility trace.
 * @param epsilon [out] Component of eligibility vector corresponding to the adaptive firing threshold variable.
 * @param weight [in, out] Synaptic weight.
 * @param cp [in] Common properties for synapses.
 * @param optimizer [in] Instance of weight optimizer.
 *
 */


 /**
 * Compute gradient change for eprop synapses.
 *
 * This method is called from an eprop synapse on the eprop target neuron. It updates various parameters related to e-prop plasticity
 * according to Bellec et al. (2020).
 *
 * @param presyn_isis [in, out] Vector of inter-spike intervals.
 * @param t_previous_update [in] Time of the last update.
 * @param t_previous_trigger_spike [in] Time of the last trigger spike.
 * @param kappa [in] Decay factor for the eligibility trace.
 * @param average_gradient [in] Boolean flag determining whether to compute an average of the gradients over the given period.
 *
 * @return Returns the computed gradient value.
 */
 
 ```

Reply:
@JesusEV This looks quite good to me, except that it seems counterintuitive that `z_previous_buffer` is also an output variable, given that it is a value from a previous time step. Could you add a short explanation?

Reply:
@heplesser When calculating the gradient between spikes, it will occasionally happen that the 2 spikes will happen one right after the other (with a separation of 1 time step). Given the way the function is built, `z_previous_buffer` will carry forward a value of 1 for the following integration if this happens. Most of the time, that variable just contains a zero, as this is what we have between separate spikes. I will look into incorporating this information into this documentation block.

Reply:
I am still a bit confused here and wondering about more elegant solutions. We can have three cases (I am looking at the code of `eprop_iaf_psc_delta`, I think the others are the same):

1. In a corner case, we could have more than one spike at the same time, i.e., `t_spike == t_previous`. This can happen in practice if a `parrot_neuron` forwards the output of a high-rate spike generator representing input. In the current implementation, the loop in `compute_gradient()` would not be executed at all until a spike with `t_spike > t_previous` arrives. Is this a problem? This could also be avoided by making the input synapses non-plastic.
2. We can have `t_spike == t_previous + 1`. In this case, the loop will be exeucted once and in that, and only in that case, will we have `z_previous_buffer == 1` when the function returns.
3. If `t_spike > t_previous + 1`, the function will return with `z_previous_buffer == 0`.

Now `z` can only take on values 0 or 1, and this value has an effect only on the line changing `z_bar`. 
1. If we enter the function with `z_previous_buffer == 0`, then `z` will be 1 only during the second round through the loop, corresponding to `t = t_previous_spike + 1`. At this time, `z_bar` gets a "kick".
2. If we enter the function with `z_previous_buffer == 1`, then `z` will be 1 during the first and second round through the loop, so `z_bar` gets kicked at `t_previous_spike` and `t_previous_spike+1`.

I think I see now how this works if we have two spikes right after each other. Let's assume spikes at 1, 5, 6, and 9 steps. Then,
- for `t_spike == 5`, we enter with `zpb == 0` and `t_p == 1`, so we kick `z_bar` at 2; we return `zbp==0`
- for `t_spike == 6`, we enter with `zpb == 0` and `t_p == 5`, so we do not kick and return with `zpb==1`
- for `t_spike == 9`, we enter with `zpb == 1` and `t_p == 6`, and thus kick at 6 and 7, making up for the missed kick when processing the previous spike.

But what would happen if we had spikes at 1, 5, 6, 7 and 9? Then, we would not kick `z_bar` at all when calling for `t_spike == 6` and `7` (because in neither case we go through the loop), and when calling for `9`, we would kick only at 7 and 8, but not at 6.

Now I wonder a little why `z_bar` needs to be updated first from `t_prev` to `t_prev+1` (and gradients/weights updated) before `z_bar` is "kicked" due to the new spike. If one could "kick" right at `t_prev`, things would get easier. If we need to wait until `t_prev+1`, could we do the `+ V_.P_z_in_ * z` inside the loop on the second round through the loop, and add it after the loop if there is only a single round through the loop? This should solved the problem with 5, 6, 7, although solving 5, 5, 5 might need some more thought. But we would not need the `z_{current,previous}_buffer`.

Reply:
After in person discussion, I think I got it now, code looks correct at least as long as there is at most one spike per time step. And since we compute the gradient here for a single synapse, there is no plausible way in which more than one spike could pass per time step, so we don't need to consider that corner case. Maybe one could add an `assert( t_spike > t_previous_spike );`?